core/shared/src/main/scala/magnolia.scala

```diff
@@ -505,15 +507,14 @@ object Magnolia {
     *
     *  This method is intended to be called only from code generated by the Magnolia macro, and
     *  should not be called directly from users' code. */
-  def subtype[Tc[_], T, S <: T](name: String, tc: => Tc[S], isType: T => Boolean, asType: T => S) =
-    new Subtype[Tc, T] {
+  def subtype[Tc[_], T, S <: T](name: String, tc: => Tc[S], isType: T => Boolean, asType: T => S): Subtype[Tc, T] =
+    new Subtype[Tc, T] with PartialFunction[T, S] {
       type SType = S
       def label: String = name
       def typeclass: Tc[SType] = tc
-      def cast: PartialFunction[T, SType] = new PartialFunction[T, S] {
-        def isDefinedAt(t: T) = isType(t)
-        def apply(t: T): SType = asType(t)
-      }
+      def cast: PartialFunction[T, SType] = this
```

Comment:
I will remove this, as we're still pre-1.0, and I'm not offering any API guarantees.